### ISSUE TITLE
Give a more obvious warning if a database is opened read-only, especially when adding/editing

### ISSUE THREAD

**Issue by tremby** [keepassxreboot/keepassxc#7123]
## Summary
It is not obvious enough when a database is opened read-only. Entries can be added and edited and there is very little indication that the changes are not being saved.

## Examples
- Unless the user has specifically asked to open the database read-only (though I don't see an option), give an obtrusive warning saying it was opened read-only
- Either disallow editing entries, or give an obvious warning when an edit is performed
- Either disallow creating new entries, or give an obvious warning when an entry is added, or when the add option is used
- Either disallow deleting entries, or give an obvious warning when a deletion is performed

## Context
I recently updated my OS and found the recommended way of installing is now via a snap. It now doesn't at all integrate with my theme and that may be why I didn't notice the subtle "[Read-only]" in the titlebar. But it seems that presumably because of some kind of security thing, keepassxc is opening the database file as read-only. I had no idea, and over the last week or so probably made various edits to the database. I found today that one I definitely added just earlier is not there. So I've had about a week's worth of data loss just because I didn't know the database wasn't saving.

**Comment by droidmonkey** [comment]
We have no concept of "read only". If the database fails to save after a change then the save as dialog will pop up requesting for a writable location to be selected.

**Comment by tremby** [comment]
> We have no concept of "read only".

Are you sure? Then what clues it in to display "[Read-only]" in the
window title?

> If the database fails to save after a change then the save as dialog will pop up requesting for a writable location to be selected.

This is absolutely not happening for me, on version 2.4.3.

There is no relevant output on the console either. I did not find an
option to start in any kind of verbose or debug mode.

**Comment by droidmonkey** [comment]
We are on version 2.6.6

**Comment by tremby** [comment]
Ah. I see what has happened.

I first installed from snap, since I saw this was the recommended way. I quickly found that I could not open my passwords database -- the snap can't access the NFS mount it's in. I was probably in a hurry to get something done, and I then switched to the older version in the Ubuntu apt repository, then forgot I had done that. Today `snap refresh` was telling me everything was up to date so I thought I was on the latest.

I see now that if I install the latest version the title bar does not say "[Read-only]" but then there's a red banner if I try to save something to an unwriteable file.

Now I need to research how to allow this snap to read files from my NFS share.

**Comment by droidmonkey** [comment]
https://keepassxc.org/docs/#faq-appsnap-homedir

**Comment by tremby** [comment]
It was inside the home directory, but pointing outside via a symlink.
Mounting it with --bind instead of using a symlink did the job.

**Comment by phoerious** [comment]
Where did you read that snap was the recommended way?

**Comment by tremby** [comment]
Good question. Maybe I'm having multiple brain farts. I do see you have an
official Ubuntu PPA. I think possibly when trying to figure out how to get
a recent version of some other package I was told the recommended way was
now with snaps rather than PPAs, and I may have misremembered it as being
recommended by keepassxc. Chromium now wants to be installed via snap,
maybe that's what influenced me?

It does seem nicer to have somewhere central rather than a whole pile of
PPAs such I need to update each time I do a release upgrade. I know because
of the encapsulation dependency hell is going to be less common. The
sandboxing is probably a good thing for security. And I hear it's easier
for the software authors and distribution maintainers, since they don't
need to maintain so many packages. Those are my understandings of the
benefits, but there seem to be some drawbacks too -- theme integration
seems to be an unsolved problem, libraries are duplicated, the sandboxing
can cause annoyances like I had earlier, maybe others?

I'd be interested to hear your take.

Sorry for the noise though. This whole ticket was me being an idiot.

**Comment by phoerious** [comment]
Canonical is pushing Snaps, so they will probably the future unless they kill it off again and adopt Flatpak. Snaps are supposed to make things simpler but as of now, they are more like a nightmare for us and make everything more complicated due to their half-assed encapsulation mechanism.

**Comment by MitchellHicks** [comment]
"Sorry for the noise though. This whole ticket was me being an idiot."

This is me everyday...

> We have no concept of "read only". If the database fails to save after a change then the save as dialog will pop up requesting for a writable location to be selected.

does this mean when you open the database you can not tell the filesystem has it marked "READ ONLY"?

I have a version that is marked read only and shared to computers i keep forgetting i am using the read only version. so some warning etc would be nice... 

Great work.

**Comment by droidmonkey** [comment]
We used to detect this, but it caused more issues than it was worth (people editing entries and not realizing their changes were not saved). That is why we moved to this model. If you want to keep a read only copy of a database, it would be better to create a primary database that has a folder setup with keeshare in import only mode. This allows you to make changes to the read only entries without impacting the source database.
